File: Grapevine/Cpp/code/src/groupDefinition.cpp
#include <cassert>

#include "groupUtils.h"
#include "util.h"
#include "groupDefinition.h"

void GroupDefinition::handleContextSummary(GroupContextSummary& currentGroupSummary, const ContextSummary& newSummary)
{
    assert(&currentGroupSummary != NULL);
    assert(&newSummary != NULL);
    
    int uid = newSummary.getId();
    int gid = currentGroupSummary.getId();
    
    vector<int> groupIds;
    GroupUtils::getDeclaredMemberships(newSummary, groupIds);
    vector<int> memberIds;
    currentGroupSummary.getMemberIds(memberIds);
    if (Util::in(groupIds, gid)) 
    {
        if (!Util::in(memberIds, uid)) // uid not in 
            currentGroupSummary.addMemberId(uid);
    }
}

void GroupDefinition::handleGroupSummary(GroupContextSummary& currentGroupSummary, const GroupContextSummary& newSummary)
{
    vector<int> memberIds;
    currentGroupSummary.getMemberIds(memberIds);
    vector<int> newMemberIds;
    newSummary.getMemberIds(newMemberIds);
    
    Util::removeAll(newMemberIds, memberIds); // the newMemberIds holds the result
    currentGroupSummary.addMemberIds(newMemberIds);
}

void GroupDefinition::handleContextSummary(GroupContextSummary* currentGroupSummary, const ContextSummary* newSummary)
{
    return handleContextSummary(*currentGroupSummary, *newSummary);
}

void GroupDefinition::handleGroupSummary(GroupContextSummary* currentGroupSummary, const GroupContextSummary* newSummary)
{
    return handleContextSummary(*currentGroupSummary, *newSummary);
}

File: Grapevine/Python/src/groupDefinition.py
from util.groupUtils import *
from util.util import *
from groupContextSummary import *

class GroupDefinition(object):
    def __init__(self, gId):
        self.gId = gId
    
    def handleContextSummary(self, currentGroupSummary, newSummary):
        assert(type(currentGroupSummary) is GroupContextSummary)
        assert(type(newSummary) is ContextSummary)
        """
        Make sure currentGroupSummary has newSummary in its member
        when newSummary has currentGroupSummary as its group
        """
        uid = newSummary.getId()
        gid = currentGroupSummary.getId()
        groupIds = getDeclaredMemberships(newSummary)
        
        #print currentGroupSummary
        if gid in groupIds:
            if uid not in currentGroupSummary.getMemberIds():
                currentGroupSummary.addMemberId(uid)
        #print currentGroupSummary
    
    def handleGroupSummary(self, currentGroupSummary, newGroupSummary):
        assert(type(currentGroupSummary) is GroupContextSummary)
        assert(type(newGroupSummary) is GroupContextSummary)
        memberIds = currentGroupSummary.getMemberIds()
        newMemberIds = newGroupSummary.getMemberIds()
        
        # ??? The new
        newMemberIds = removeAll(newMemberIds, memberIds)
        #print newMemberIds
        currentGroupSummary.addMemberIds(newMemberIds)
        
    def getId(self):
        return self.gId
    
if __name__ == "__main__":
    import sys
    sys.path.append("../test")
    from testGroupDefinition import *
    unittest.main(verbosity=2)
